primitives/src/circuit/merkle_tree/sparse_merkle_tree.rs

```diff
@@ -0,0 +1,688 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Circuit implementation of a sparse, 3-ary Merkle tree, instantiated
+//! with a Rescue hash function.
+
+use crate::{
+    circuit::rescue::RescueNativeGadget,
+    merkle_tree::{
+        internal::MerkleNode, prelude::RescueSparseMerkleTree, MerkleTreeScheme, ToTraversalPath,
+    },
+    rescue::RescueParameter,
+};
+use ark_std::vec::Vec;
+use jf_relation::{errors::CircuitError, BoolVar, Circuit, PlonkCircuit, Variable};
+
+type SparseMerkleTree<F> = RescueSparseMerkleTree<BigUint, F>;
+
+type NodeVal<F> = <SparseMerkleTree<F> as MerkleTreeScheme>::NodeValue;
+type MembershipProof<F> = <SparseMerkleTree<F> as MerkleTreeScheme>::MembershipProof;
+type Element<F> = <SparseMerkleTree<F> as MerkleTreeScheme>::Element;
+type Index<F> = <SparseMerkleTree<F> as MerkleTreeScheme>::Index;
+use num_bigint::BigUint;
+use typenum::U3;
+
+use super::{
+    MembershipProofBooleanEncoding, MerkleNodeBooleanEncoding, MerkleTreeGadget,
+    MerkleTreeHelperGadget, Rescue3AryNodeVar, SparseMerkleTreeGadget, StandardLeafVar,
+};
+
+#[derive(Debug, Clone)]
+/// Circuit variable for a Merkle authentication path for a Rescue-based, 3-ary
+/// Merkle tree.
+pub struct Rescue3AryMembershipProofVar {
```

Comment:
This struct can be more general as it doesn't use any Rescue-related info inside. Is it better to call it `Merkle3AryMembershipProofVar`?

Reply:
Right, I've renamed as suggested, and also unified across the two MT circuit impls: https://github.com/EspressoSystems/jellyfish/pull/172/commits/981a20e40dfb0dcd4ef1d6f4deefd726d5114c9a.